File: src/model_story/parsers/accelerate.py
from __future__ import annotations

import json
import re

from model_story.models import RawMetric
from model_story.parsers.base import ParserContext
from model_story.parsers.registry import register_parser

# Accelerate uses HuggingFace Trainer under the hood but may also emit
# step-based logs via `accelerate.logging` with different keys.
# Primary signal: JSON dicts with 'loss' or 'eval_loss' and a 'step' key.
_ACCEL_DICT = re.compile(r"^\s*\{['\"](?:loss|eval_loss)['\"].*['\"]step['\"]")


@register_parser
class AccelerateParser:
    name = "accelerate"
    priority = 78

    def sniff(self, sample_lines: list[str]) -> float:
        hits = sum(1 for line in sample_lines if _ACCEL_DICT.match(line))
        return min(hits / max(len(sample_lines), 1) * 5, 0.88)

    def parse_line(self, line: str, ctx: ParserContext) -> list[RawMetric]:
        stripped = line.strip()
        if not stripped.startswith("{"):
            return []

        normalized = stripped.replace("'", '"').replace("True", "true").replace("False", "false")
        try:
            data: dict[str, object] = json.loads(normalized)
        except json.JSONDecodeError:
            return []

        step = data.pop("step", None)
        if isinstance(step, (int, float)):
            ctx.current_step = int(step)

        epoch = data.pop("epoch", None)
        if isinstance(epoch, (int, float)):
            ctx.current_epoch = float(epoch)

        metrics: list[RawMetric] = []
        for key, val in data.items():
            if not isinstance(val, (int, float)):
                continue
            metrics.append(
                RawMetric(
                    seq=ctx.seq,
                    epoch=ctx.current_epoch,
                    step=ctx.current_step,
                    key=key,
                    value=float(val),
                    parser_name=self.name,
                    confidence=0.85,
                )
            )
        return metrics

File: epochix-vscode/src/parsers/accelerate.ts
/**
 * TypeScript port of src/epochix/parsers/accelerate.py.
 *
 * `accelerator.print({...})` dicts keyed by step: {'loss': 0.61, 'step': 100}.
 */
import type { Parser, ParserContext, RawMetric } from "./base";

const ACCEL_DICT = /^\s*\{['"](?:loss|eval_loss)['"].*['"]step['"]/;

export class AccelerateParser implements Parser {
  readonly name = "accelerate";
  readonly priority = 78;

  sniff(sampleLines: readonly string[]): number {
    const hits = sampleLines.filter((l) => ACCEL_DICT.test(l)).length;
    return Math.min((hits / Math.max(sampleLines.length, 1)) * 5, 0.88);
  }

  parseLine(line: string, ctx: ParserContext): RawMetric[] {
    const stripped = line.trim();
    if (!stripped.startsWith("{")) return [];
    const normalized = stripped
      .replace(/'/g, '"')
      .replace(/True/g, "true")
      .replace(/False/g, "false");
    let data: Record<string, unknown>;
    try {
      data = JSON.parse(normalized) as Record<string, unknown>;
    } catch {
      return [];
    }
    if (data === null || typeof data !== "object") return [];
    const step = data["step"];
    delete data["step"];
    if (typeof step === "number") ctx.currentStep = Math.trunc(step);
    const epoch = data["epoch"];
    delete data["epoch"];
    if (typeof epoch === "number") ctx.currentEpoch = epoch;

    const metrics: RawMetric[] = [];
    for (const [key, val] of Object.entries(data)) {
      if (typeof val !== "number") continue;
      metrics.push({
        seq: ctx.seq,
        epoch: ctx.currentEpoch,
        step: ctx.currentStep,
        key,
        value: val,
        parserName: this.name,
        confidence: 0.85,
      });
    }
    return metrics;
  }
}
